### PR TITLE
Added try_recv & try_send to UnixDatagram

### PR DESCRIPTION
**Pull request by jean-airoldie**
This allows nonblocking sync send & recv operations on the socket.

### Motivation
This is useful for draining the socket's buffer before dropping it. To do so, one would shutdown the writing end of the peer than `try_recv` until it errors.

### CODE REVIEW

**Comment by carllerche** [comment]
This seems fine to me. Looks like Ci is failing.

Thoughts @tokio-rs/maintainers ?

**Comment by jean-airoldie** [comment]
Seems like freebsd doesn't return WouldBlock for some reason.

**Comment by jean-airoldie** [comment]
Turns out that freebsd return `Os { code: 55, kind: Other, message: "No buffer space available" }'` instead of `WouldBlock` when the datagram queue is full.

**Comment by Ralith** [comment]
> This is useful for draining the socket's buffer before dropping it.

When is that needed?

File: tokio-net/src/uds/datagram.rs [hunk 1]
```diff
@@ -75,6 +75,11 @@ impl UnixDatagram {
         poll_fn(|cx| self.poll_send_priv(cx, buf)).await
     }
 
+    /// Try to send a datagram from the socket without blocking.
+    pub fn try_send(&mut self, buf: &[u8]) -> io::Result<usize> {
+        self.io.get_mut().send(buf)
```
Comment (kleimkuhler) [inline tokio-net/src/uds/datagram.rs]:
```suggestion
        self.io.get_ref().send(buf)
```

Should be okay here!

**Comment by jean-airoldie** [comment]
> When is that needed?

I'm passing shared memory objects so I need to properly deallocate them when dropping the channel.

**Comment by carllerche** [comment]
I'm good w/ this. Could we fix the merge conflicts and I would :heart: if the API docs could be padded out a bit and have an example.

**Comment by jean-airoldie** [comment]
Almost forgot about this PR.

I added the `try_send_to` and `try_recv_from` methods as well. I improved the documentation and added 2 basics tests in the doc to show usage.

**Comment by kleimkuhler** [comment]
@jean-airoldie Sorry there has not been any new activity on this. It should still be good to go, but it looks like `rusfmt` is failing. The build was cleared, but it looks like `try_send_to` needs a comma after `P: AsRef<Path>`.

Once the build is passing again it should be 👍

**Comment by jean-airoldie** [comment]
I'll rebase against master to see if this fixes it. If I run rustfmt on my machine I see nothing wrong.

**Comment by jean-airoldie** [comment]
I manually edited it. Its weird that I doesn't show up on my machine because I'm using the same rustfmt version than the CI.

**Comment by kleimkuhler** [comment]
@jean-airoldie When manually checking formatting I usually do `cargo fmt --all -- --check`, which on your branch before the previous commit did *not* show any errors. If that is how you were checking, I don't think an error shows up because the `datagram.rs` file is not included in the default features.

CI seems to address this by passes all the files manually with a `find` command.

**Comment by Darksonn** [comment]
This would also make sense to add to the split `UnixDatagram`. However, I don't think it should block merging this PR.

**Pull request merged**